### ISSUE TITLE
iPhone X - Space above keyboard

### ISSUE THREAD

**Issue by ezescaruli** [Papercloud/KeyboardConstraint#10]
I've tried `KeyboardConstraint` in an iPhone X and it's having an issue when the keyboard appears. It looks like this (note the space between the sign up button and the keyboard):

![image](https://user-images.githubusercontent.com/5781314/35248959-34f3eb38-0024-11e8-9303-09eef075a61c.png)

I debugged the point in which the keyboard appears:
https://github.com/Papercloud/KeyboardConstraint/blob/bcd2a3e9da89cd39d74332784583e348aef0c84a/KeyboardConstraint/KeyboardAdjustConstraint.m#L65
and had this:
```
(lldb) po [self firstItem]
<UILayoutGuide: 0x6000005acf60 - "UIViewSafeAreaLayoutGuide", layoutFrame = {{0, 0}, {375, 690}}, owningView = <UIView: 0x7fb63fc06df0; frame = (0 88; 375 724); autoresize = W+H; layer = <CALayer: 0x600000024b80>>>
```
and then I did:
```
(lldb) po [[self firstItem] conformsToProtocol:@protocol(UILayoutSupport)]
error: cannot find protocol declaration for 'UILayoutSupport'
```
The debugger doesn't find the `UILayoutSupport` protocol declaration, which is _weird_...

In runtime, what happens is that the above condition resolves to `NO`, so the execution never falls in the line that decrements `keyboardHeight`, and that's why we still have that extra space in the screenshot.

I've tried replacing the fix with this, and it seems to be working fine:
```objc
- (void)_keyboardDidChangeVisible:(NSNotification *)notification {
    ...

    if (notification.name == UIKeyboardWillShowNotification) {
        if (@available(iOS 11.0, *)) {
            UILayoutGuide *layoutGuide = [self layoutGuideItem];
            
            if (layoutGuide != nil) {
                keyboardHeight -= layoutGuide.owningView.safeAreaInsets.bottom;
            }
        }
    }

    ...
}

- (UILayoutGuide *)layoutGuideItem {
    if ([self.firstItem isKindOfClass:[UILayoutGuide class]]) {
        return self.firstItem;
    }
    
    if ([self.secondItem isKindOfClass:[UILayoutGuide class]]) {
        return self.secondItem;
    }
    
    return nil;
}
```

**Comment by ezescaruli** [comment]
@markst What do you reckon? I know that the fix that we merged the other day worked for you, but for some reason it didn't work when I tried it (both in a simulator and a device), so I replaced it with the code I posted above to make it work.

But maybe I'm doing something wrong...?

**Comment by markst** [comment]
Yo @ezescaruli. Thanks for the concise issue report. 
We were aware of this when supporting iPhone X, see comments: https://github.com/Papercloud/KeyboardConstraint/pull/9#issuecomment-358483288

Are you sure you're using the `bottomLayoutGuide` when attaching the keyboard constraint?
Since the bottom layout guide resolves to the safe area insets on an iPhone X. 

The `layoutGuide.length` is considered when calculating the keyboard height:
https://github.com/Papercloud/KeyboardConstraint/blob/master/KeyboardConstraint/KeyboardAdjustConstraint.m#L67

**Comment by ezescaruli** [comment]
@markst So, the sign up button bottom constraint looks like this:

![image](https://user-images.githubusercontent.com/5781314/35252673-a81970cc-0035-11e8-94ee-70d5c40c8ffe.png)

> The layoutGuide.length is considered when calculating the keyboard height

Yep that's right, however the issue in my case is that the execution is not reaching that line because both
```objc
[self.firstItem conformsToProtocol:@protocol(UILayoutSupport)]
```
and
```objc
[self.secondItem conformsToProtocol:@protocol(UILayoutSupport)]
```
resolve to `NO`. I _think_ the reason is that the safe area layout guides are instances of `UILayoutGuide`, but don't conform to the `UILayoutSupport` protocol.

**Comment by markst** [comment]
@ezescaruli as mentioned you need to use the `bottomLayoutGuide` as example:

![image](https://user-images.githubusercontent.com/274318/35252826-5be89e0c-0036-11e8-843c-8016d01bb956.png)

**Comment by ezescaruli** [comment]
@markst I don't have the bottom layout guide among the options 🤔 

![image](https://user-images.githubusercontent.com/5781314/35252895-b28e3bfe-0036-11e8-88b4-ba4ef0ac2feb.png)

I could probably make it appear if I disable `Use Safe Area Layout Guides` in the storyboard file inspector, but that would not be ideal, would it?

**Comment by markst** [comment]
Yup you need to disable safe area guides in the interface builder settings:
![image](https://user-images.githubusercontent.com/274318/35252974-1159572c-0037-11e8-80ff-90382e8e2f12.png)

By all means have a crack at getting it to work fine for safe area insets.
Just be sure you don't break it for `bottomLayoutGuide`.

**Comment by ezescaruli** [comment]
@markst Perfecto, I'll give a try and check if the fix I made above works for the bottom layout guide too 👍
